### PR TITLE
Hook implementations

### PR DESCRIPTION
**Pull request by laserkelvin**
# ALCHEMI Toolkit Pull Request

## Description

This PR provides concrete implementations of a preliminary set of generally useful hooks for use in the dynamics.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

- NaN detection hook, which will flag trajectories that result in NaN for either energies or forces
- ProfilerHook, which will provide CUDA event or perf_counter timings at each dynamics stage
- BiasPotentialHook, allows users to plug in arbitrary changes to energy and forces during dynamics
- WrapPeriodicHook, re-wraps atomic positions
- SnapshotHook, which is the de facto trajectory logging mechanism. Saves batches with a thread pool to make it async.

## Testing

- [ ] Unit tests pass locally (`make pytest`)
- [ ] Linting passes (`make lint`)
- [ ] New tests added for new functionality meets coverage expectations?

## Checklist

- [ ] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [ ] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/dynamics/hooks/_utils.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        return out
+
+    # For amax, amin, mean — use scatter_reduce_
+    fill = {
+        "amax": float("-inf"),
+        "amin": float("inf"),
+        "mean": 0.0,
+    }[reduce]
+    out = torch.full((num_graphs,), fill, device=values.device, dtype=values.dtype)
+    out.scatter_reduce_(0, batch_idx, values, reduce=reduce, include_self=False)
```
Comment (dallasfoster) [inline nvalchemi/dynamics/hooks/_utils.py]: We might have warp utilities for this, if it is worth the hassle.

File: nvalchemi/dynamics/hooks/_utils.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    Float[Tensor, "B 1"]
+        Kinetic energy per graph.
+    """
+    # Ensure masses is (V,) for element-wise multiply
+    m = masses.squeeze(-1) if masses.dim() > 1 else masses
+    # KE per atom: 0.5 * m * ||v||^2
+    ke_per_atom = 0.5 * m * (velocities * velocities).sum(dim=-1)  # (V,)
+    # Sum per graph
+    ke = scatter_reduce_per_graph(ke_per_atom, batch_idx, num_graphs, reduce="sum")
+    return ke.unsqueeze(-1)  # (B, 1)
```
Comment (dallasfoster) [inline nvalchemi/dynamics/hooks/_utils.py]: This might exist in toolkit-ops too.

File: nvalchemi/dynamics/hooks/_utils.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    Returns
+    -------
+    Float[Tensor, "B"]
+        Instantaneous kinetic temperature per graph in Kelvin.
+    """
+    ke = kinetic_energy_per_graph(velocities, masses, batch_idx, num_graphs).squeeze(
+        -1
+    )  # (B,)
+    n_atoms = atoms_per_graph.float()  # (B,)
+    return (2.0 * ke) / (3.0 * n_atoms * KB_EV)
```
Comment (dallasfoster) [inline nvalchemi/dynamics/hooks/_utils.py:328]: Same as before.

File: nvalchemi/dynamics/hooks/_utils.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    inv_cell = torch.linalg.inv(per_atom_cell)  # (V, 3, 3)
+    frac = torch.einsum("vi,vij->vj", positions, inv_cell)  # (V, 3)
+
+    # Wrap only periodic dimensions
+    wrapped_frac = torch.where(per_atom_pbc, frac % 1.0, frac)  # (V, 3)
+
+    # Back to Cartesian: pos_i = frac_j * cell_ji
+    wrapped_pos = torch.einsum("vj,vji->vi", wrapped_frac, per_atom_cell)  # (V, 3)
+
+    return wrapped_pos
```
Comment (dallasfoster) [inline nvalchemi/dynamics/hooks/_utils.py]: O.o

File: nvalchemi/dynamics/hooks/safety.py [hunk 5]
```diff
@@ -126,10 +134,48 @@ def __call__(self, batch: Batch, dynamics: BaseDynamics) -> None:
...
+
+            bad_fields.append(key)
+            non_finite_mask = ~torch.isfinite(tensor)
+            n_bad = non_finite_mask.sum().item()
+
+            # Map back to graph indices
+            if tensor.shape[0] == batch.num_nodes:
+                # Node-level tensor: find which atoms have non-finite values
+                affected_nodes = non_finite_mask.any(dim=-1)  # (V,)
+                affected_graphs = batch.batch[affected_nodes].unique().tolist()
```
Comment (dallasfoster) [inline nvalchemi/dynamics/hooks/safety.py]: Eeek, surely there is a way we can rewrite this to avoid the .item(). Not sure I love how this hook is implemented.
